### PR TITLE
Ce 3107 restrict mw ns move

### PR DESCRIPTION
**Pull request by kamilkoterba**
https://wikia-inc.atlassian.net/browse/CE-3107

Moving into within and out of MW namespace for whitelisted pages was allowed.
This is restricting moves for whitelisted pages as well.

Also did some refactoring of 'move' error messages to be more verbose and display allowed groups as well and in same way changed check for 'undelete' to check for 'editinterfacetrusted' right and reuse message for 'undelete' that already exists.

@Wikia/community-engineering

### CODE REVIEW

File: includes/wikia/Wikia.php [hunk 1]
```diff
@@ -30,6 +30,8 @@
 $wgHooks['AjaxAddScript']            [] = 'Wikia::onAjaxAddScript';
 $wgHooks['TitleGetSquidURLs']        [] = 'Wikia::onTitleGetSquidURLs';
 $wgHooks['userCan']                  [] = 'Wikia::canEditInterfaceWhitelist';
+$wgHooks['getUserPermissionsErrors'] [] = 'Wikia::canMoveMediaWikiNS';
+$wgHooks['getUserPermissionsErrors'] [] = 'Wikia::canUndeleteMediaWikiNS';
```
Comment (lgarczewski) [inline includes/wikia/Wikia.php]: These two could be handled in one hook since the code of the hook functions is nearly identical.

File: includes/Exception.php [hunk 2]
```diff
@@ -374,26 +374,33 @@ class PermissionsError extends ErrorPageError {
...
 		}
 
 		$this->errors = $errors;
 	}
 
+	/**
+	 * @param string $permission Name of action
+	 * @return array error message key and message params (optionally)
+	 */
+	public static function prepareBadAccessErrorArray( $permission ) {
```
Comment (lgarczewski) [inline includes/Exception.php:390]: Why has this been moved out of the constructor? Did I miss this being used somewhere else? Seems to still just be used in one place -- but then what is the purpose of this change?

File: includes/specials/SpecialUndelete.php [hunk 3]
```diff
@@ -328,9 +328,9 @@ function isDeleted() {
 	function undelete( $timestamps, $comment = '', $fileVersions = array(), $unsuppress = false ) {
 		global $wgUser;
 
-		$resultMock = '';
-		if ( !$this->title->userCan( 'undelete' ) ) {
```
Comment (lgarczewski) [inline includes/specials/SpecialUndelete.php:332]: Could you explain why the change from Title::userCan() to Title::getPermissionsErrors()? Could we not have used the hook in Title::userCan like we do with Wikia::canEditInterfaceWhitelist?

Comment (kamilkoterba) [inline includes/specials/SpecialUndelete.php:332]: getPermissionsErrors allows us to pass errors returned from hooks to user, while userCan calls same logic but maps result to boolean saying is there where any errors.
So in general if we wany to just check permissions we chould use userCan but if we want to throw it we should check using getPermissionsErrors

File: includes/Exception.php (hunk 2 above)
Comment (adamkarminski) [inline includes/Exception.php:390]: It will allow us reusing just this bit without duplicating the logic. I think that in the context of all changes this one makes sense.

File: includes/wikia/Wikia.php (hunk 1 above)
Comment (adamkarminski) [inline includes/wikia/Wikia.php]: Agreed.

File: includes/specials/SpecialUndelete.php (hunk 3 above)
Comment (adamkarminski) [inline includes/specials/SpecialUndelete.php:332]: I'm with Kamil on this one. This is a very strange and bad UX not to give users a reason for which you blocked an action for them.

File: includes/wikia/Wikia.php [hunk 4]
```diff
@@ -2331,16 +2333,26 @@ static function onLocalFilePurgeThumbnailsUrls( LocalFile $file, Array &$urls )
 	 * usually return true to allow processing other hooks
 	 * return false stops permissions processing and we are totally decided (nothing later can override)
 	 */
-	static function canEditInterfaceWhitelist (&$title, &$wgUser, $action, &$result) {
+	static function canEditInterfaceWhitelist ( &$title, &$wgUser, $action, &$result ) {
 		global $wgEditInterfaceWhitelist;
 
+		// Allowed actions at this point
+		$notRelevantActions = [
```
Comment (adamkarminski) [inline includes/wikia/Wikia.php]: This name is **very** misleading

File: includes/wikia/Wikia.php [hunk 5]
```diff
@@ -2331,16 +2333,26 @@ static function onLocalFilePurgeThumbnailsUrls( LocalFile $file, Array &$urls )
...
+			'read',
+			'move', // Is being checked in next hook canMoveMediaWikiNS
+			'undelete' // Is being checked in next hook canUndeleteMediaWikiNS
+		];
+
 		// List the conditions we don't care about for early exit
-		if ( $action == "read" || $title->getNamespace() != NS_MEDIAWIKI || empty( $wgEditInterfaceWhitelist )) {
+		if ( in_array( $action, $notRelevantActions )
+			|| $title->getNamespace() != NS_MEDIAWIKI
```
Comment (adamkarminski) [inline includes/wikia/Wikia.php:2347]: `!$title->inNamespace( NS_MEDIAWIKI )`

**Comment by adamkarminski** [comment]
Address the comments and it will LGTM. :)
